bytecomp/bytegen.ml

```diff
@@ -34,20 +34,53 @@ let new_label () =
 (**** Operations on compilation environments. ****)
 
 let empty_env =
-  { ce_stack = Ident.empty; ce_heap = Ident.empty; ce_rec = Ident.empty }
+  { ce_stack = Ident.empty; ce_closure = Not_in_closure }
 
 (* Add a stack-allocated variable *)
 
 let add_var id pos env =
   { ce_stack = Ident.add id pos env.ce_stack;
-    ce_heap = env.ce_heap;
-    ce_rec = env.ce_rec }
+    ce_closure = env.ce_closure }
 
 let rec add_vars idlist pos env =
   match idlist with
     [] -> env
   | id :: rem -> add_vars rem (pos + 1) (add_var id pos env)
 
+(* Compute the closure environment *)
+
+type function_definition =
+  | Single_non_recursive
+  | Multiple_recursive of Ident.t list
+
+let closure_entries fun_defs fvs =
+  let rec add_positions entries pos_to_entry pos delta = function
```

Comment:
`pos` and `delta` could be labelled parameters to make callsites easier to read.